plugins/legion-hid/fu-legion-hid-device.c

```diff
@@ -0,0 +1,649 @@
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "device report start command failed with %u",
+			    status);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_legion_hid_device_upgrade_query_size(FuLegionHidDevice *self,
+					guint8 id,
+					guint *max_size,
+					GError **error)
+{
+	g_autoptr(GByteArray) res = NULL;
+	g_autoptr(FuStructLegionHidUpgradeCmd) st_cmd = fu_struct_legion_hid_upgrade_cmd_new();
+	FuLegionHidUpgradeRetryHelper helper = {0};
+	guint8 status = 0;
+	guint8 content[] = {0x02,
+			    FU_LEGION_HID_UPGRADE_STEP_QUERY_SIZE,
+			    FU_LEGION_HID_CMD_CONSTANT_SN};
+
+	fu_struct_legion_hid_upgrade_cmd_set_length(st_cmd, sizeof(content) + 5);
+	fu_struct_legion_hid_upgrade_cmd_set_device_id(st_cmd, id);
+	if (!fu_struct_legion_hid_upgrade_cmd_set_data(st_cmd, content, sizeof(content), error))
+		return FALSE;
+	if (!fu_udev_device_write(FU_UDEV_DEVICE(self),
+				  st_cmd->buf->data,
+				  st_cmd->buf->len,
+				  FU_LEGION_HID_DEVICE_IO_TIMEOUT,
+				  FU_IO_CHANNEL_FLAG_NONE,
+				  error))
+		return FALSE;
+
+	res = g_byte_array_sized_new(FU_LEGION_HID_DEVICE_FW_REPORT_LENGTH);
+	helper.res = res;
+	helper.main_id = fu_struct_legion_hid_upgrade_cmd_get_main_id(st_cmd);
+	helper.sub_id = fu_struct_legion_hid_upgrade_cmd_get_sub_id(st_cmd);
+	helper.dev_id = id;
+	helper.step = FU_LEGION_HID_UPGRADE_STEP_QUERY_SIZE;
+	if (!fu_legion_hid_device_read_upgrade_response(self, &helper, error))
+		return FALSE;
+	status = res->data[9];
+	if (status == FU_LEGION_HID_RESPONSE_STATUS_FAIL) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "device report query size command failed with %u",
+			    status);
+		return FALSE;
+	}
+	if (max_size)
+		*max_size = fu_memread_uint16(res->data + 9, G_BIG_ENDIAN);
+
+	return TRUE;
+}
+
+static gboolean
+fu_legion_hid_device_upgrade_write_data(FuLegionHidDevice *self,
+					guint8 id,
+					GByteArray *buffer,
+					guint max_size,
+					GError **error)
+{
+	guint size = buffer->len;
+	guint inner_loop_times = max_size / FU_LEGION_HID_DEVICE_FW_PACKET_LENGTH;
+	guint outer_loop_times = size / max_size;
+	guint send_size = 0;
+	guint ready_send_size = 0;
+	guint recieved_size = 0;
+	g_autoptr(FuStructLegionHidUpgradeCmd) st_cmd = fu_struct_legion_hid_upgrade_cmd_new();
+	g_autoptr(GByteArray) res = g_byte_array_sized_new(FU_LEGION_HID_DEVICE_FW_REPORT_LENGTH);
+	guint8 content[FU_LEGION_HID_DEVICE_FW_PACKET_LENGTH + 1] = {0};
+	FuLegionHidUpgradeRetryHelper helper = {0};
+	gboolean wait = FALSE;
+
+	if (max_size % FU_LEGION_HID_DEVICE_FW_PACKET_LENGTH != 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "device report max size %u is invalid",
+			    max_size);
+		return FALSE;
+	}
+
+	if (size % max_size != 0)
+		outer_loop_times += 1;
+	for (guint i = 0; i < outer_loop_times; ++i) {
+		for (guint j = 0; j < inner_loop_times; ++j) {
```

Comment:
I think this should probably implemented using FuChunks -- and this function is very deeply nested -- you probably need to split out a fu_legion_hid_device_upgrade_write_data**_chunk**() or something like that.